src/server/api/user.ts

```diff
@@ -238,42 +234,64 @@ router.patch('/url/ownership', async (req, res) => {
 })
 
 /**
- * Endpoint for user to edit a longUrl.
+ * Endpoint for user to edit a file or a longUrl.
+ *
+ * If editing a file link, only the file can be changed and not
+ * the long URL. This is to ensure a one-to-one mapping between
+ * the short URL and S3 object key.
  */
-router.patch('/url/edit', validateUrls, async (req, res) => {
-  const { userId, longUrl, shortUrl } = req.body
-  try {
-    const user = await User.scope({
-      method: ['includeShortUrl', shortUrl],
-    }).findOne({
-      where: { id: userId },
-    })
+router.patch(
+  '/url/edit',
+  fileUploadMiddleware,
+  validateUrls,
+  async (req, res) => {
+    const { userId, longUrl, shortUrl } = req.body
+    // @ts-ignore Type definition does not know about the compulsory file field.
+    const file: UploadedFile | undefined = req.files?.file
+    try {
+      const user = await User.scope({
+        method: ['includeShortUrl', shortUrl],
+      }).findOne({
+        where: { id: userId },
+      })
 
-    if (!user) {
-      res.notFound(jsonMessage('User not found.'))
-      return
-    }
+      if (!user) {
+        res.notFound(jsonMessage('User not found.'))
+        return
+      }
 
-    const [url] = (user.get() as UserType).Urls
+      const [url] = (user.get() as UserType).Urls
 
-    if (!url) {
-      res.notFound(jsonMessage(`Short link "${shortUrl}" not found for user.`))
-    }
+      if (!url) {
+        res.notFound(
+          jsonMessage(`Short link "${shortUrl}" not found for user.`),
+        )
+      }
 
-    await transaction((t) => url.update({ longUrl }, { transaction: t }))
-    res.ok(jsonMessage(`Short link "${shortUrl}" has been updated`))
+      await transaction(async (t) => {
+        if (!url.isFile) {
+          await url.update({ longUrl }, { transaction: t })
+        } else if (file) {
+          const { uploadFileToS3 } = container.get<S3Interface>(
```

Comment:
same here

Reply:
`container.get()` should be called with the imports at the top of file as resolving dependencies can be an expensive op